admin_confirm/tests/unit/test_admin_options.py

```diff
@@ -0,0 +1,312 @@
+
+from admin_confirm.constants import CACHE_KEYS
+
+
+@mock.patch.object(ShoppingMallAdmin, "inlines", [])
+class TestAdminOptions(AdminConfirmTestCase):
+    @mock.patch.object(ShoppingMallAdmin, "confirmation_fields", ["name"])
+    @mock.patch.object(ShoppingMallAdmin, "fields", ["name", "town"])
+    def test_m2m_field_not_in_fields(self):
+        gm = GeneralManager.objects.create(name="gm")
+        shops = [ShopFactory() for i in range(3)]
+        town = Town.objects.create(name="town")
+        mall = ShoppingMall.objects.create(name="mall", general_manager=gm, town=town)
+        mall.shops.set(shops)
+
+        # new values
+        gm2 = GeneralManager.objects.create(name="gm2")
+        shops2 = [ShopFactory() for i in range(3)]
+        town2 = Town.objects.create(name="town2")
+
+        data = {
+            "id": mall.id,
+            "name": "name",
+            "town": town2.id,
+            "_confirm_change": True,
+            "_continue": True,
+        }
+        response = self.client.post(
+            f"/admin/market/shoppingmall/{mall.id}/change/", data=data
+        )
+
+        # Should be shown confirmation page
+        self._assertSubmitHtml(
+            rendered_content=response.rendered_content, save_action="_continue"
+        )
+
+        # Should not have cached the unsaved obj
+        cached_item = cache.get(CACHE_KEYS["object"])
+        self.assertIsNone(cached_item)
+
+        # Should not have saved changes yet
+        self.assertEqual(ShoppingMall.objects.count(), 1)
+        mall.refresh_from_db()
+        self.assertEqual(mall.name, "mall")
+        self.assertEqual(mall.general_manager, gm)
+        self.assertEqual(mall.town, town)
+        for shop in mall.shops.all():
+            self.assertIn(shop, shops)
+
+        # Click "Yes, I'm Sure"
+        confirmation_received_data = data
+        del confirmation_received_data["_confirm_change"]
+        confirmation_received_data["_confirmation_received"] = True
+
+        response = self.client.post(
+            f"/admin/market/shoppingmall/{mall.id}/change/",
+            data=confirmation_received_data,
+        )
+
+        # Should not have redirected to changelist
+        self.assertEqual(response.url, f"/admin/market/shoppingmall/{mall.id}/change/")
+
+        # Should have saved obj
+        self.assertEqual(ShoppingMall.objects.count(), 1)
+        saved_item = ShoppingMall.objects.all().first()
+        # should have updated fields that were in form
+        self.assertEqual(saved_item.name, data["name"])
+        self.assertEqual(saved_item.town, town2)
+        # should have presevered the fields that are not in form
+        self.assertEqual(saved_item.general_manager, gm)
+        for shop in saved_item.shops.all():
+            self.assertIn(shop, shops)
+
+        # Should have cleared cache
+        for key in CACHE_KEYS.values():
+            self.assertIsNone(cache.get(key))
```

Comment:
Maybe 🤔 create a _assert helper method for this. It's required in most tests, but... it's bad practice to have any assert which is not explicit in the test func body.